[Xuyan0518]iP

Xuyan0518 opened the pull request:
# **Friday**
> *"In a little time I began to speak to him and teach him to speak to me; and first, I made him know his Name should be Friday, which was the Day I saved his Life; I called him so for the Memory of the Time." -- Robinson Crusoe ([source](https://bookroo.com/quotes/friday))* 

## Overview
Friday is a digital secretary designed to assist users with various tasks, including task management, scheduling, and providing information on specific topics.

## It is:
- [x] **Free** to use.
- [x] User-friendly.
- [x] Text-based.

## Features
- Task Management: add, delete, mark tasks as done, and view the task list.
- Scheduling: Add events with start and end times.
- Information Retrieval: Retrieve information based on user queries.
- Data Persistence: Store task and event data for future sessions.

## How to install: 
1. download it from [here](https://github.com/Xuyan0518/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :grin:

## Here is the `main` method:
```ruby
public static void main(String[] args) {
        new Friday().run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Xuyan-0518 committed:
Level-0

Xuyan-0518 committed:
Level-0

Xuyan-0518 committed:
Merge remote-tracking branch 'origin/master'

Xuyan-0518 committed:
Level_0

Xuyan-0518 committed:
Level_1

Xuyan-0518 committed:
Level_2

Xuyan-0518 committed:
Level_2

Xuyan-0518 committed:
Level_2

Xuyan-0518 committed:
Level-5

Xuyan-0518 committed:
Level-5

Xuyan-0518 committed:
Level-6

Xuyan-0518 committed:
A-TextUiTesting

Xuyan-0518 committed:
Level-7

Xuyan-0518 committed:
Level-7

Xuyan-0518 committed:
Level-8

Xuyan-0518 committed:
Merge branch 'branch-level-8'

Xuyan-0518 committed:
A-MoreOOP

Xuyan-0518 committed:
A-Packages

Xuyan-0518 committed:
A-Packages

Xuyan-0518 committed:
Merge branch 'add-gradle-support'

Xuyan-0518 committed:
A-Gradle

Xuyan-0518 committed:
A-Junit

Xuyan-0518 committed:
A-Jar

Xuyan-0518 committed:
A-JavaDoc

Xuyan-0518 committed:
Add Javadoc for all packages in friday

Briefly describe the purpose and scope of the Javadoc changes. Include any additional context that may help reviewers understand the motivation behind the documentation updates.

- Specify the classes, methods, or packages for which Javadoc was added.
- Include any significant details about the information added, such as parameters, return values, or exceptions.

Xuyan-0518 committed:
Add find functionality

Introduce a new feature allowing users to search for tasks by entering keywords. The `searchTask` method in the `TaskList` class checks for matching tasks, and the `containsKeyword` method in the `Task` class facilitates the search process. Enhances user experience by providing a convenient way to find tasks based on specific criteria.

pjanthony2001 commented on src/main/java/friday/main/Friday.java:
```diff
@@ -0,0 +1,137 @@
+ */
+public class Friday {
+    private static final String DATA_FILE_PATH = "./src/main/java/data/Friday.txt";
+    private static final String HORIZONTAL_LINE = "_".repeat(60);
```
Good that the constants are named correctly!

Xuyan0518 replied:
thx!

pjanthony2001 commented on src/main/java/friday/storage/Storage.java:
```diff
@@ -0,0 +1,95 @@
+     * @throws IOException If an I/O error occurs during file appending.
+     */
+    public void appendToFile(String textToAppend) throws IOException {
+        FileWriter fw = new FileWriter(filePath, true);
```
Maybe fw could be written as fileWriter instead for clarity?

Xuyan0518 replied:
okay, will change!

getsquared commented on src/main/java/friday/task/TaskList.java:
```diff
@@ -0,0 +1,254 @@
+     *
+     * @return The number of tasks in the task list.
+     */
+    public int length() {
```
Perhaps the name of this method could be a verb like getLength()?

Xuyan0518 replied:
Thx! Will change.

getsquared commented on src/main/java/friday/main/Friday.java:
```diff
@@ -0,0 +1,137 @@
+ * Tasks are stored in a data file, and users can interact with Friday by entering commands.
+ */
+public class Friday {
+    private static final String DATA_FILE_PATH = "./src/main/java/data/Friday.txt";
```
I like that your constant names are in uppercase with underscores 👍

Xuyan0518 replied:
Thx!

getsquared commented on src/main/java/friday/task/Task.java:
```diff
@@ -0,0 +1,116 @@
+ */
+public class Task {
+    protected String description;
+    protected boolean isDone;
```
All your booleans are named to sound like booleans, great job 👍

Xuyan0518 replied:
Thx!

getsquared commented on src/main/java/friday/task/Task.java:
```diff
@@ -0,0 +1,116 @@
+    public static Task parseTask(String taskString) {
+        String cat = taskString.substring(1, 2);
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM dd yyyy HHmm");
+        switch (cat) {
```
Your switch statements are indented correctly 👍

Xuyan0518 replied:
Thx!

getsquared commented on src/main/java/friday/storage/Storage.java:
```diff
@@ -0,0 +1,95 @@
+        try (Scanner fileScanner = new Scanner(file)) {
+            while (fileScanner.hasNextLine()) {
+                String taskData = fileScanner.nextLine();
+                if (!taskData.isEmpty()) {
```
Could have guard clauses instead, to prevent nesting too much?

Xuyan0518 replied:
Thx! Will improve!

getsquared reviewed (commented):
LGTM 👍 Just some minor changes.

Xuyan-0518 committed:
Implement checkstyle

Changes made in this commit:
- Added checkstyle configuration file.
- Integrated checkstyle checks into the build process.
- Addressed existing code issues identified by checkstyle.

Xuyan-0518 committed:
Implemented GUI functionality

 Added JavaFX components for user interaction, including text input field, send button, and dialog container for displaying messages. Integrated user input processing with backend logic to add, mark, delete, and search tasks. Implemented error handling and exception messages for robust user experience.

pjanthony2001 commented on src/main/java/friday/main/Friday.java:
```diff
@@ -0,0 +1,183 @@
+    private static final String DATA_FILE_PATH = "./src/main/java/data/Friday.txt";
+
+    /** The user interface handler for displaying messages and interacting with users. */
+    private final Ui ui;
+
+    /** The storage handler for reading from and writing to the data file. */
+    private final Storage storage;
```
Good that your UI and Storage fields are set to final here!

Xuyan0518 replied:
thx!

pjanthony2001 commented on src/main/java/friday/parser/Parser.java:
```diff
@@ -0,0 +1,82 @@
+                sb.append(i + 1 + ". " + tasks.getTask(i).toString());
+                sb.append(System.lineSeparator());
+            }
+            return sb.toString();
+        case "mark":
+            String responseToMark = tasks.markTask(userInput);
+            try {
+                storage.writeToFile(tasks);
+            } catch (IOException e) {
+                return e.getMessage();
+            }
+            return responseToMark;
+        case "unmark":
+            String responseToUnmark = tasks.unmarkTask(userInput);
+            try {
+                storage.writeToFile(tasks);
+            } catch (IOException e) {
+                return e.getMessage();
+            }
+            return responseToUnmark;
+        case "todo":
+            return tasks.addTodo(userInput);
+        case "deadline":
+            return tasks.addDeadline(userInput);
+        case "event":
+            return tasks.addEvent(userInput);
+        case "delete":
+            String responseToDelete = tasks.deleteTask(userInput);
+            try {
+                storage.writeToFile(tasks);
+            } catch (IOException err) {
+                return err.getMessage();
+            }
+            return responseToDelete;
+        case "find":
+            return tasks.searchTask(userInput);
+        case "bye":
+            return "Bye Master. Have a nice day!";
+        default:
+            return "HUH? What do you mean?";
```
Nice that your switch case is indented properly!

Xuyan0518 replied:
thx!

pjanthony2001 commented on src/main/java/friday/parser/Parser.java:
```diff
@@ -0,0 +1,82 @@
+ * It interacts with the TaskList and Storage classes to manage tasks and data storage.
+ */
+public class Parser {
+    private TaskList tasks;
+
+    private Storage storage;
```
Maybe you could also set some of these to be final as well if needed.

Xuyan0518 replied:
alright! will change!

pjanthony2001 commented on src/main/java/friday/task/Event.java:
```diff
@@ -0,0 +1,76 @@
+ * Subclass of the Task class.
+ */
+public class Event extends Task {
+    protected LocalDateTime from;
+    protected LocalDateTime to;
+
```
Should these fields be protected or private? Maybe setting them to private for their visibility might be better.

Xuyan0518 replied:
okay! will improve!

pjanthony2001 reviewed (commented):
Overall, really good, no big issues. I think most of the small issues arise as artifacts from previous iterations. Try to just scan through your code one time to catch these small errors. Good job!

Xuyan-0518 committed:
This commit updates the program's entry point to Launcher.java instead of Friday.java

By doing this, we ensure that the application launches through the Launcher class,
which is responsible for setting up the JavaFX application correctly.

This change is necessary to improve the organization of the codebase and adhere
to best practices for structuring JavaFX applications.

No functional changes were made to the application logic.

Xuyan-0518 committed:
Add assertions for input validation and boundary checks

This commit introduces assertions to ensure proper arguments are passed around. Assertions have been added to validate user input, such as ensuring it is not null, and to verify that task indices are within the bounds of the task list. These assertions enhance the robustness and reliability of the task management functionality, aiding in debugging and preventing potential runtime errors.

Xuyan-0518 committed:
Follow code quality

It increases readability of the code by following the code quality.

Xuyan-0518 committed:
Implement C-MassOps features for delete, mark and unmark commands

It is useful to implement mass ops for these commands because users can delete/mark/unmark all the tasks that they want in one go. So users no longer need to type out the operation each time when they are trying to perform the same operation on multiple tasks.

Xuyan-0518 committed:
Implement C-MassOps features for delete, mark and unmark commands

It is useful to implement mass ops for these commands because users can delete/mark/unmark all the tasks that they want in one go. So users no longer need to type out the operation each time when they are trying to perform the same operation on multiple tasks.

Xuyan0518 committed:
Merge pull request #1 from Xuyan0518/branch-A-Assertions

Add assertions for input validation and boundary checks

Xuyan-0518 committed:
Update Ui.png

Ui.png shows what the main page of the chatbot looks like.

Xuyan-0518 committed:
Update User Guide

User guide provides guidance to users on how to use this application.

Xuyan-0518 committed:
Update .gitignore

Friday.txt file is moved to .gitignore so that it will not cause conflicts for future git commits.

Xuyan-0518 committed:
Update README.md

Update the format of README.md so that it follows the correct format.

Xuyan-0518 committed:
Update README.md

Update the format of README.md so that it follows the correct format.

Xuyan0518 committed:
Ui.png

Xuyan-0518 committed:
Bug fixes

Make sure task list in the parser class is not null.

Xuyan-0518 committed:
Bug fixes

Make sure task list in the parser class is not null.

Xuyan-0518 committed:
Update file path

Original file path has issues after packaging the program into jar file. So file path is changed to solve this situation.

Xuyan0518 committed:
Delete docs/Ui.png

Xuyan-0518 committed:
Merge remote-tracking branch 'origin/master'

Xuyan0518 committed:
Add files via upload
